### ISSUE TITLE
Add per-rule reportDisables secondary option property

### ISSUE THREAD

**Issue by nex3** [stylelint/stylelint#4875]
> What is the problem you're trying to solve?

Within Google, we want different levels of enforcement for different lints. For lints that can produce false positives, we want users to be able to opt out of them by writing `stylelint-disable` comments, but for lints that we're highly confident are always the right thing to do we don't want anyone opting out.

> What solution would you like to see?

I'd like to see an `ignoreDisables` option as a per-rule configuration that controls whether that rule in particular can be disabled.

I can work around this by running stylelint multiple times with multiple different configurations, but that's more complex and less efficient than being able to configure it rule-by-rule.

**Comment by malsf21** [comment]
Hey! Sorry, it looks like we've been a bit slow at triaging. Thank you for using the issue template, and this seems like a reasonable request to me - definitely more convenient than running stylelint multiple times with different configs (that all need to be interpreted independently!). I'll tag it as "Needs Discussion" just to loop in another member of the org.

**Comment by jeddy3** [comment]
@nex3 Thank you for the request. And as @malsf21 said, apologies for the slow triaging; it seems like we've all been busy of late.

This is an interesting use case. I've been working through how we might go about achieving it. I think an option that causes stylelint to ignore certain `stylelint-disable` commands will have a problematic user experience. The commands will be silently ignored and it won't be obvious to a user why that is.

We could add a `reportDisallowedDisables` option and a per-rule `disallowDisables` secondary option property. The linter would then error with a clear message if a `stylelint-disable` command is scoped to one of these rules. It'll be in line with the other `report*Disables` options, including the proposed [`reportDescriptionlessDisables`](https://github.com/stylelint/stylelint/issues/4868) and [`reportUnscopedDisables`](https://github.com/stylelint/stylelint/issues/2292) ones.

```json
{
  "rules": {
    "color-no-invalid-hex": [true, { "disallowDisables": true }]
  }
}
```

```css
/* stylelint-disable color-no-invalid-hex */
a { color: #aa }
```

```console
1:1 error Disallowed disabled "color-no-invalid-hex"
```

Would that approach cater to your use case? It'd only be effective in conjunction with the `reportUnscopedDisables` option, but it sounds like you'll be wanting to turn on that too for your level of enforcement.

---

Going forward, I think we'll want to [support these top-level configuration options like `--report*` and `--fix` in the configuration object](https://github.com/stylelint/stylelint/issues/4843#issuecomment-650137005), rather than just on the command line.

**Comment by m-allanson** [comment]
I had a thought about this the other day. It’s half-baked so feel free to shut it down or ignore it as appropriate. 

Could this behaviour be handled by a custom reporter? Allow stylelint to report disables as normal, then let the reporter decide which ones to report on.

I haven’t thought about how config would work, or whether it sets a confusing precedent about the location of stylelint functionality.

On Sun, 9 Aug 2020, at 16:23, Richard Hallows wrote:
> 

> @nex3 <https://github.com/nex3> Thank you for the request. And as @malsf21 <https://github.com/malsf21> said, apologies for the slow triaging; it seems like we've all been busy of late.

> This is an interesting use case. I've been working through how we might go about achieving it. I think an option that causes stylelint to ignore certain `stylelint-disable` commands will have a problematic user experience. The commands will be silently ignored and it won't be obvious to a user why that is.

> We could add a `reportDisallowedDisables` option and a per-rule `disallowDisables` secondary option property. The linter would then error with a clear message if a `stylelint-disable` command is scoped to one of these rules. It'll be in line with the other `report*Disables` options, including the proposed `reportDescriptionlessDisables` <https://github.com/stylelint/stylelint/issues/4868> and `reportUnscopedDisables` <https://github.com/stylelint/stylelint/issues/2292> ones.

> {
  "rules": {
    "color-no-invalid-hex": [true, { "disallowDisables": true }]
  }
}

**Comment by nex3** [comment]
> Would that approach cater to your use case? It'd only be effective in conjunction with the `reportUnscopedDisables` option, but it sounds like you'll be wanting to turn on that too for your level of enforcement.

@jeddy3 Yeah, that would work well for our use-case.

> Could this behaviour be handled by a custom reporter? Allow stylelint to report disables as normal, then let the reporter decide which ones to report on.

@m-allanson This would require reporters to be able to control how a given message affects the exit code of the program, which seems potentially out of a reporter's scope.

**Comment by nex3** [comment]
@jeddy3 Thinking about this further, does it make sense to have a top-level `reportDisallowedDisables` option? Or should any rules with `disallowDisables: true` automatically be reported? In the former case, you can get into the weird state where disallowed disables aren't reported but disables are still listed as disallowed, which has unclear semantics.

**Comment by jeddy3** [comment]
> Thinking about this further, does it make sense to have a top-level `reportDisallowedDisables` option?

I had the same thought. It probably doesn't make sense.

Perhaps we call the per-rule option `reportDisables: true` to make it clear to the user that this option is within the same fold as the (top-level) `report*Disables` options:

- Existing:
  - `reportNeedlessDisables`
  - `reportInvalidScopeDisables`
- Proposed:
  - `reportDescriptionlessDisables`
  - `reportUnscopedDisables `

I think we can mark this feature as "ready to implement" as we can, if needed, revisit the option name once we've a pull request on the go.
